File: src/com/hkb48/keepdo/DatabaseAdapter.java
package com.hkb48.keepdo;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import android.annotation.TargetApi;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.os.Build;
import android.os.Environment;
import android.util.Log;

import com.hkb48.keepdo.Database.TaskCompletion;
import com.hkb48.keepdo.Database.TasksToday;

class DatabaseAdapter {
    private static final String TAG = "#KEEPDO_DB_ADAPTER: ";
    private static final String SDF_PATTERN_YMD = "yyyy-MM-dd"; 
    private static final String SDF_PATTERN_YM = "yyyy-MM";
    private static final String SELECT_FORM = "select * from ";
    private static final String SELECT_ARG_FORM = " where ";

    private static DatabaseAdapter INSTANCE = null;
    private DatabaseHelper mDatabaseHelper = null;
    private SQLiteDatabase mDatabase = null;

    mDatabaseHelper = DatabaseHelper.getInstance(context.getApplicationContext());
}

    static synchronized DatabaseAdapter getInstance(Context context) {
        if (INSTANCE == null) {
            INSTANCE = new DatabaseAdapter(context);


    private DatabaseAdapter(Context context) {
    	}

    	return INSTANCE;
    }

    private SQLiteDatabase openDatabase() {
    	synchronized (this) {
	    	try {
	    		mDatabase = mDatabaseHelper.getWritableDatabase();
	    	} catch (SQLiteException e) {
                if (BuildConfig.DEBUG) {
                    Log.d(TAG, e.getLocalizedMessage());
                }
	    		throw e;
	    	}
    	}
 
    	return mDatabase;
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    private void closeDatabase() {
    	synchronized (this) {
			if (mDatabase != null) {
				mDatabase.close();
				mDatabase = null;
			}
    	}
    }

    public synchronized void close() {
		mDatabaseHelper.close();
    }

    public List<Task> getTaskList() {
        List<Task> tasks = new ArrayList<Task>();
        String selectQuery = SELECT_FORM + TasksToday.TABLE_NAME
                + " order by " + TasksToday.TASK_LIST_ORDER + " asc;";

        Cursor cursor = openDatabase().rawQuery(selectQuery, null);

        if (cursor.moveToFirst()) {
            do {
                tasks.add(getTask(cursor));
            } while (cursor.moveToNext());
        }

        cursor.close();
        closeDatabase();

        return tasks;
    }

    public void addTask(Task task) {
        long rowID = Task.INVALID_TASKID;
        String taskName = task.getName();
        String taskContext = task.getContext();
        Recurrence recurrence = task.getRecurrence();
        Reminder reminder = task.getReminder();

        if ((taskName == null) || (taskName.isEmpty()) || (recurrence == null) || (reminder == null)) {
            return;
        }

        try {
            ContentValues contentValues = new ContentValues();
            contentValues.put(TasksToday.TASK_NAME, taskName);
            contentValues.put(TasksToday.FREQUENCY_MON, String.valueOf(recurrence.getMonday()));
            contentValues.put(TasksToday.FREQUENCY_TUE, String.valueOf(recurrence.getTuesday()));
            contentValues.put(TasksToday.FREQUENCY_WEN, String.valueOf(recurrence.getWednesday()));
            contentValues.put(TasksToday.FREQUENCY_THR, String.valueOf(recurrence.getThurday()));
            contentValues.put(TasksToday.FREQUENCY_FRI, String.valueOf(recurrence.getFriday()));
            contentValues.put(TasksToday.FREQUENCY_SAT, String.valueOf(recurrence.getSaturday()));
            contentValues.put(TasksToday.FREQUENCY_SUN, String.valueOf(recurrence.getSunday()));
            contentValues.put(TasksToday.TASK_CONTEXT, taskContext);
            contentValues.put(TasksToday.REMINDER_ENABLED, String.valueOf(reminder.getEnabled()));
            contentValues.put(TasksToday.REMINDER_TIME, String.valueOf(reminder.getTimeInMillis()));
            contentValues.put(TasksToday.TASK_LIST_ORDER, task.getOrder());

            rowID = openDatabase().insertOrThrow(TasksToday.TABLE_NAME, null, contentValues);
            closeDatabase();
            
        } catch (SQLiteException e) {
            Log.e(TAG, e.getMessage());
        }
    }

    void editTask(Task task) {
        Long taskID = task.getTaskID();
        String taskName = task.getName();
        String taskContext = task.getContext();
        Recurrence recurrence = task.getRecurrence();
        Reminder reminder = task.getReminder();

        if ((taskName == null) || (taskName.isEmpty()) || (recurrence == null) || (reminder == null)) {
            return;
        }

        ContentValues contentValues = new ContentValues();
        contentValues.put(TasksToday.TASK_NAME, taskName);
        contentValues.put(TasksToday.FREQUENCY_MON, String.valueOf(recurrence.getMonday()));
        contentValues.put(TasksToday.FREQUENCY_TUE, String.valueOf(recurrence.getTuesday()));
        contentValues.put(TasksToday.FREQUENCY_WEN, String.valueOf(recurrence.getWednesday()));
        contentValues.put(TasksToday.FREQUENCY_THR, String.valueOf(recurrence.getThurday()));
        contentValues.put(TasksToday.FREQUENCY_FRI, String.valueOf(recurrence.getFriday()));
        contentValues.put(TasksToday.FREQUENCY_SAT, String.valueOf(recurrence.getSaturday()));
        contentValues.put(TasksToday.FREQUENCY_SUN, String.valueOf(recurrence.getSunday()));
        contentValues.put(TasksToday.TASK_CONTEXT, taskContext);
        contentValues.put(TasksToday.REMINDER_ENABLED, String.valueOf(reminder.getEnabled()));
        contentValues.put(TasksToday.REMINDER_TIME, String.valueOf(reminder.getTimeInMillis()));
        contentValues.put(TasksToday.TASK_LIST_ORDER, task.getOrder());

        String whereClause = TasksToday._ID + "=?";
        String whereArgs[] = {taskID.toString()};

        try {
        	openDatabase().update(TasksToday.TABLE_NAME, contentValues, whereClause, whereArgs);
        	closeDatabase();
        } catch (SQLiteException e) {
            Log.e(TAG, e.getMessage());
        }
    }

    void deleteTask(Long taskID) {
        // Delete task from TASKS_TABLE_NAME
    	String whereClause = TasksToday._ID + "=?";
        String whereArgs[] = {taskID.toString()};
        openDatabase().delete(TasksToday.TABLE_NAME, whereClause, whereArgs);
        closeDatabase();

        // Delete records of deleted task from TASK_COMPLETION_TABLE_NAME
        whereClause = TaskCompletion.TASK_NAME_ID + "=?";
        openDatabase().delete(TaskCompletion.TABLE_NAME, whereClause, whereArgs);
        closeDatabase();
    }

	void setDoneStatus(Long taskID, Date date, Boolean doneSwitch) {
        if (taskID == null) {
            return;
        }

        SimpleDateFormat dateFormat = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);

        if (doneSwitch != null && doneSwitch) {
            ContentValues contentValues = new ContentValues();

            contentValues.put(TaskCompletion.TASK_NAME_ID, taskID);
            contentValues.put(TaskCompletion.TASK_COMPLETION_DATE, dateFormat.format(date));

            try {
            	openDatabase().insertOrThrow(TaskCompletion.TABLE_NAME, null, contentValues);
            } catch (SQLiteException e) {
                Log.e(TAG, e.getMessage());
            } finally {
            	closeDatabase();
            }

        } else {
            String whereClause = TaskCompletion.TASK_NAME_ID + "=? and " + TaskCompletion.TASK_COMPLETION_DATE + "=?";
            String whereArgs[] = {taskID.toString(), dateFormat.format(date)};
            openDatabase().delete(TaskCompletion.TABLE_NAME, whereClause, whereArgs);
            closeDatabase();
        }
    }

    boolean getDoneStatus(Long taskID, Date day) {
        boolean isDone = false;
        String selectQuery = SELECT_FORM + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?";
        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        SimpleDateFormat sdf_ymd = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);

        if (cursor.moveToFirst()) {
            do {
                Date date = getDate(cursor);
                if (date != null) {
                    if (sdf_ymd.format(date).equals(sdf_ymd.format(day))) {
                        isDone = true;
                        break;
                    }
                }
            } while (cursor.moveToNext());
        }

        cursor.close();
        closeDatabase();

        return isDone;
    }

    public int getNumberOfDone(Long taskID) {
        int numberOfDone = 0;
        String selectQuery = SELECT_FORM + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?";
        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        if (cursor != null) {
            numberOfDone = cursor.getCount();
            cursor.close();
        }
        closeDatabase();
        return numberOfDone;
    }

    public Date getFirstDoneDate(Long taskID) {
        final SimpleDateFormat sdf_ymd = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);
        Date date = null;
        String selectQuery =  "select min(" + TaskCompletion.TASK_COMPLETION_DATE + ") from "
                + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?";
        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        if (cursor != null) {
            cursor.moveToFirst();
            String dateString = cursor.getString(0);
            if (dateString != null) {
                try {
                    date = sdf_ymd.parse(dateString);
                } catch (ParseException e) {
                    Log.e(TAG, e.getMessage());
                }
            }
            cursor.close();
        }
        closeDatabase();
        return date;
    }

    public Date getLastDoneDate(Long taskID) {
        final SimpleDateFormat sdf_ymd = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);
        Date date = null;
        String selectQuery =  "select max(" + TaskCompletion.TASK_COMPLETION_DATE + ") from "
                + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?";
        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        if (cursor != null) {
            cursor.moveToFirst();
            String dateString = cursor.getString(0);
            if (dateString != null) {
                try {
                    date = sdf_ymd.parse(dateString);
                } catch (ParseException e) {
                    Log.e(TAG, e.getMessage());
                }
            }
            cursor.close();
        }
        closeDatabase();
        return date;
    }

    Task getTask(Long taskID) {
        Task task = null;
        String selectQuery = SELECT_FORM + TasksToday.TABLE_NAME + SELECT_ARG_FORM + TasksToday._ID + "=?";

        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        if (cursor != null){
            cursor.moveToFirst();
            task = getTask(cursor);
            cursor.close();
        }

        closeDatabase();

        return task;
    }

    ArrayList<Date> getHistory(Long taskID, Date month) {
        ArrayList<Date> dateList = new ArrayList<Date>();
        String selectQuery = SELECT_FORM + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?";

        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});
        SimpleDateFormat sdf_ym = new SimpleDateFormat(SDF_PATTERN_YM, Locale.JAPAN);

        if (cursor.moveToFirst()) {
            do {
                Date date = getDate(cursor);
                if (date != null) {
                    if (sdf_ym.format(date).equals(sdf_ym.format(month))) {
                        dateList.add(date);
                    }
                }
            } while (cursor.moveToNext());
        }

        cursor.close();
        closeDatabase();
        
        return dateList;
    }

    ComboCount getComboCount(Long taskID) {
    	int currentCount = 0;
    	int maxCount = 0;

        // Task Recurrence
        Recurrence recurrence = getTask(taskID).getRecurrence();

    	// execute SQL to get task completion dates
        String selectQuery = SELECT_FORM + TaskCompletion.TABLE_NAME + SELECT_ARG_FORM + TaskCompletion.TASK_NAME_ID + "=?" 
        					+ " order by " + TaskCompletion.TASK_COMPLETION_DATE + " asc;";
        Cursor cursor = openDatabase().rawQuery(selectQuery, new String[] {String.valueOf(taskID)});

        // Count
        if (cursor != null) {
        	if (cursor.moveToFirst()) {
                Calendar calToday = getCalendar(DateChangeTimeUtil.getDate());
        		Calendar calDone = getCalendar(getDate(cursor));
        		Calendar calIndex = (Calendar)calDone.clone();
        		boolean isCompleted = false;
            	do {
            		if (calIndex.equals(calDone)) {
            			// count up combo
            			currentCount ++;
        				if (currentCount > maxCount) {
            				maxCount = currentCount;
        				}
            			if (cursor.moveToNext()) {
            				calDone = getCalendar(getDate(cursor));
            			} else {
            				isCompleted = true;
            			}
            			calIndex.add(Calendar.DAY_OF_MONTH, 1);
            		} else {
            			if (recurrence.isValidDay(calIndex.get(Calendar.DAY_OF_WEEK))) {
            				// stop combo
            				if (!calIndex.equals(calToday)) {
            					currentCount = 0;
            				}
            				if (!isCompleted) {
            					calIndex = (Calendar)calDone.clone();
            				} else {
                    			calIndex.add(Calendar.DAY_OF_MONTH, 1);
            				}
            			} else {
                			calIndex.add(Calendar.DAY_OF_MONTH, 1);
            			}
            		}
            	} while (!calIndex.after(calToday));
        	}
        	cursor.close();
        }

        closeDatabase();

        return new ComboCount(currentCount, maxCount);
    }

    private Date getDate(Cursor cursor) {
        String dateString = cursor.getString(cursor.getColumnIndex(TaskCompletion.TASK_COMPLETION_DATE));
        SimpleDateFormat sdf_ymd = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);
        Date date = null;
        if (dateString != null) {
            try {
                date = sdf_ymd.parse(dateString);
            } catch (ParseException e) {
                Log.e(TAG, e.getMessage());
            }
        }
        return date;
    }

    private Calendar getCalendar(Date date) {
    	Calendar calendar = Calendar.getInstance();
    	calendar.setTime(date);
    	return calendar;
    }

    private Task getTask(Cursor cursor) {
        Long taskID = Long.parseLong(cursor.getString(cursor.getColumnIndex(TasksToday._ID)));
        String taskName = cursor.getString(cursor.getColumnIndex(TasksToday.TASK_NAME));
        String taskContext = cursor.getString(cursor.getColumnIndex(TasksToday.TASK_CONTEXT));

        Recurrence recurrence = new Recurrence(Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_MON))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_TUE))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_WEN))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_THR))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_FRI))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_SAT))),
                                               Boolean.valueOf(cursor.getString(cursor.getColumnIndex(TasksToday.FREQUENCY_SUN))));

        Reminder reminder;
        String reminderEnabled = cursor.getString(cursor.getColumnIndex(TasksToday.REMINDER_ENABLED));
        String reminderTime = cursor.getString(cursor.getColumnIndex(TasksToday.REMINDER_TIME));
        long taskOrder = cursor.getInt(cursor.getColumnIndex(TasksToday.TASK_LIST_ORDER));
        if ((reminderEnabled == null) || (reminderTime == null)) {
            reminder = new Reminder();
        } else {
            reminder = new Reminder(Boolean.valueOf(reminderEnabled), Long.valueOf(reminderTime));
        }

        Task task = new Task(taskName, taskContext, recurrence);
        task.setReminder(reminder);
        task.setTaskID(taskID);
        task.setOrder(taskOrder);
        return task;
    }

    public int getMaxSortOrderId() {
        int maxOrderId = 0;
        String selectQuery = "select max(" + TasksToday.TASK_LIST_ORDER +") from " + TasksToday.TABLE_NAME;
        Cursor cursor = openDatabase().rawQuery(selectQuery, null);
        if (cursor != null) {
            cursor.moveToFirst();
            String idString = cursor.getString(0);
            if (idString != null) {
                maxOrderId = Integer.parseInt(idString);
            }
            cursor.close();
        }
        closeDatabase();

        return maxOrderId;
    }

    /**
     * Backup and Restore database.
     * @param backupDirPath
     * @param backupFileName
     */
    void backupDataBase(String backupDirPath, String backupFileName) {
    	File dir = new File(backupDirPath);

    	if (!dir.exists()) {
            if (!dir.mkdir()) {
                if (BuildConfig.DEBUG) {
                    Log.d(TAG, "backDataBase failed.");
                }
            }
        }

    	final String DB_FILE_PATH = mDatabaseHelper.databasePath();
		copyDataBase(DB_FILE_PATH, backupDirPath + backupFileName);
    }

    void restoreDataBase(String backupPath) {
    	final String DB_FILE_PATH = mDatabaseHelper.databasePath();
    	copyDataBase(backupPath, DB_FILE_PATH);
    }

    private synchronized void copyDataBase(String fromPath, String toPath) {
		InputStream inputStream = null;
		OutputStream outputStream = null;

		try {
			inputStream = new FileInputStream(fromPath);
			outputStream = new FileOutputStream(toPath);
			
	        byte[] buffer = new byte[1024];
	        int length;

			while ((length = inputStream.read(buffer)) > 0) {
			    outputStream.write(buffer, 0, length);
			}    	        
	        outputStream.flush();

		} catch (IOException e) {
            if (BuildConfig.DEBUG) {
            	Log.e(TAG, e.getMessage());
            }
		} finally {
			try {
    	         if (outputStream != null) {
    	        	 outputStream.close();
    	         }

    	         if (inputStream != null) {
    	        	 inputStream.close();
    	         }
			} catch (IOException e) {
                if (BuildConfig.DEBUG) {
                	Log.e(TAG, e.getMessage());
                }
			}
		}
	}
}

File: app/src/main/java/com/hkb48/keepdo/DatabaseAdapter.kt
package com.hkb48.keepdo

import android.content.ContentResolver
import android.content.Context
import android.database.sqlite.SQLiteException
import android.net.Uri
import androidx.sqlite.db.SimpleSQLiteQuery
import com.hkb48.keepdo.data.Task
import com.hkb48.keepdo.data.TaskCompletion
import com.hkb48.keepdo.data.TaskDatabase
import java.io.*
import java.util.*

class DatabaseAdapter private constructor(context: Context) {
    private val mTaskDatabase = TaskDatabase.getInstance(context)
    private val mContentResolver: ContentResolver = context.contentResolver

    val taskInfoList: MutableList<TaskInfo>
        get() {
            val taskInfoList: MutableList<TaskInfo> = mutableListOf()
            val taskList = mTaskDatabase.taskDao().getTaskListByOrder()
            for (task in taskList) {
                taskInfoList.add(getTaskInfo(task))
            }
            return taskInfoList
        }

    fun addTask(taskInfo: TaskInfo): Int {
        val taskName = taskInfo.name
        val taskContext = taskInfo.context
        val recurrence = taskInfo.recurrence
        val reminder = taskInfo.reminder
        return if (taskName == null || taskName.isEmpty()) {
            TaskInfo.INVALID_TASKID
        } else try {
            val task = Task(
                null, taskName,
                recurrence.monday,
                recurrence.tuesday,
                recurrence.wednesday,
                recurrence.thursday,
                recurrence.friday,
                recurrence.saturday,
                recurrence.sunday,
                taskContext,
                reminder.enabled,
                reminder.timeInMillis,
                taskInfo.order
            )
            mTaskDatabase.taskDao().add(task).toInt()
        } catch (e: Exception) {
            e.printStackTrace()
            TaskInfo.INVALID_TASKID
        }
    }

    fun editTask(taskInfo: TaskInfo) {
        val taskId = taskInfo.taskId
        val taskName = taskInfo.name
        val taskContext = taskInfo.context
        val recurrence = taskInfo.recurrence
        val reminder = taskInfo.reminder
        if (taskName == null || taskName.isEmpty()) {
            return
        }
        val task = Task(
            taskId, taskName,
            recurrence.monday,
            recurrence.tuesday,
            recurrence.wednesday,
            recurrence.thursday,
            recurrence.friday,
            recurrence.saturday,
            recurrence.sunday,
            taskContext,
            reminder.enabled,
            reminder.timeInMillis,
            taskInfo.order
        )
        try {
            mTaskDatabase.taskDao().update(task)
        } catch (e: SQLiteException) {
            e.printStackTrace()
        }
    }

    fun deleteTask(taskId: Int) {
        // Delete task from TASKS_TABLE_NAME
        // Records corresponding to the deleted task are also removed from TASK_COMPLETION_TABLE_NAME
        mTaskDatabase.taskDao().delete(taskId)
    }

    fun setDoneStatus(taskId: Int, date: Date, doneSwitch: Boolean) {
        try {
            if (doneSwitch) {
                val taskCompletion = TaskCompletion(null, taskId, date)
                mTaskDatabase.taskCompletionDao().insert(taskCompletion)
            } else {
                mTaskDatabase.taskCompletionDao().delete(taskId, date)
            }
        } catch (e: SQLiteException) {
            e.printStackTrace()
        }
    }

    fun getDoneStatus(taskId: Int, date: Date): Boolean {
        return mTaskDatabase.taskCompletionDao().getByDate(taskId, date).count() > 0
    }

    fun getNumberOfDone(taskId: Int): Int {
        return mTaskDatabase.taskCompletionDao().getAll(taskId).count()
    }

    fun getFirstDoneDate(taskId: Int): Date? {
        return mTaskDatabase.taskCompletionDao().getFirstCompletionDate(taskId, todayDate)
    }

    fun getLastDoneDate(taskId: Int): Date? {
        return mTaskDatabase.taskCompletionDao().getLastCompletionDate(taskId, todayDate)
    }

    fun getTask(taskId: Int): TaskInfo? {
        return mTaskDatabase.taskDao().getTask(taskId)?.let {
            getTaskInfo(it)
        }
    }

    fun getHistoryInMonth(taskId: Int, year: Int, month: Int): ArrayList<Date> {
        val calendar = Calendar.getInstance()
        calendar[Calendar.YEAR] = year
        calendar[Calendar.MONTH] = month
        val firstDayInMonth = calendar.clone() as Calendar
        firstDayInMonth[Calendar.DAY_OF_MONTH] = 1
        val lastDayInMonth = calendar.clone() as Calendar
        lastDayInMonth[Calendar.DAY_OF_MONTH] = lastDayInMonth.getActualMaximum(Calendar.DATE)

        val doneList = mTaskDatabase.taskCompletionDao().getCompletionHistoryBetween(
            taskId, firstDayInMonth.time, lastDayInMonth.time
        )
        return ArrayList(doneList)
    }

    fun getComboCount(taskId: Int): Int {
        var count = 0
        val mDoneHistory =
            mTaskDatabase.taskCompletionDao().getCompletionHistoryDesc(taskId, todayDate)
        if (mDoneHistory.isNotEmpty()) {
            val calToday = getCalendar(DateChangeTimeUtil.date)
            val calIndex = calToday.clone() as Calendar
            var calDone = getCalendar(mDoneHistory[0])
            val recurrence = getTask(taskId)!!.recurrence
            var listIndex = 0
            while (true) {
                if (calIndex == calDone) {
                    // count up combo
                    count++
                    calDone = if (++listIndex < mDoneHistory.size) {
                        getCalendar(mDoneHistory[listIndex])
                    } else {
                        break
                    }
                } else {
                    if (recurrence.isValidDay(calIndex[Calendar.DAY_OF_WEEK])) {
                        if (calIndex != calToday) {
                            break
                        }
                    }
                }
                calIndex.add(Calendar.DAY_OF_MONTH, -1)
            }
        }
        return count
    }

    fun getMaxComboCount(taskId: Int): Int {
        var currentCount = 0
        var maxCount = 0
        val mDoneHistory =
            mTaskDatabase.taskCompletionDao().getCompletionHistoryAsc(taskId, todayDate)
        if (mDoneHistory.isNotEmpty()) {
            val calToday = getCalendar(DateChangeTimeUtil.date)
            var calIndex = calToday.clone() as Calendar
            var calDone = getCalendar(mDoneHistory[0])
            val recurrence = getTask(taskId)!!.recurrence
            var listIndex = 0
            var isCompleted = false
            do {
                if (calIndex == calDone) {
                    // count up combo
                    currentCount++
                    if (currentCount > maxCount) {
                        maxCount = currentCount
                    }
                    if (++listIndex < mDoneHistory.size) {
                        calDone = getCalendar(mDoneHistory[listIndex])
                    } else {
                        isCompleted = true
                    }
                    calIndex.add(Calendar.DAY_OF_MONTH, 1)
                } else {
                    if (recurrence.isValidDay(calIndex[Calendar.DAY_OF_WEEK])) {
                        // stop combo
                        if (calIndex != calToday) {
                            currentCount = 0
                        }
                        if (isCompleted.not()) {
                            calIndex = calDone.clone() as Calendar
                        } else {
                            calIndex.add(Calendar.DAY_OF_MONTH, 1)
                        }
                    } else {
                        calIndex.add(Calendar.DAY_OF_MONTH, 1)
                    }
                }
            } while (calIndex.after(calToday).not())
        }
        return maxCount
    }

    private val todayDate: Date
        get() {
            return DateChangeTimeUtil.dateTime
        }

    private fun getCalendar(date: Date): Calendar {
        val calendar = Calendar.getInstance()
        calendar.time = date
        return calendar
    }

    private fun getTaskInfo(task: Task): TaskInfo {
        val recurrence = Recurrence(
            task.monFrequency,
            task.tueFrequency,
            task.wedFrequency,
            task.thrFrequency,
            task.friFrequency,
            task.satFrequency,
            task.sunFrequency
        )
        val taskInfo = TaskInfo(task.name, task.context, recurrence)
        taskInfo.reminder = if (task.reminderEnabled && task.reminderTime != null) {
            Reminder(task.reminderEnabled, task.reminderTime.toLong())
        } else {
            Reminder()
        }
        taskInfo.taskId = task._id ?: TaskInfo.INVALID_TASKID
        taskInfo.order = task.listOrder
        return taskInfo
    }

    val maxSortOrderId: Int
        get() {
            return mTaskDatabase.taskDao().getMaxOrder()
        }

    fun backupDataBase(outputFile: Uri): Boolean {
        var inputStream: InputStream? = null
        var outputStream: OutputStream? = null
        var success = false
        try {
            // Execute checkpoint query to ensure all of the pending transactions are applied.
            mTaskDatabase.taskDao().checkpoint((SimpleSQLiteQuery("pragma wal_checkpoint(full)")))
            mTaskDatabase.taskCompletionDao()
                .checkpoint((SimpleSQLiteQuery("pragma wal_checkpoint(full)")))

            inputStream = FileInputStream(mTaskDatabase.databasePath)
            outputStream = mContentResolver.openOutputStream(outputFile)!!
            success = copyDataBase(inputStream, outputStream)
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                outputStream?.close()
                inputStream?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        return success
    }

    fun restoreDataBase(inputFile: Uri): Boolean {
        return if (isValidSQLite(inputFile)) {
            var inputStream: InputStream? = null
            var outputStream: OutputStream? = null
            var success = false
            try {
                inputStream = mContentResolver.openInputStream(inputFile)!!
                outputStream = FileOutputStream(mTaskDatabase.databasePath)
                success = copyDataBase(inputStream, outputStream)
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                try {
                    outputStream?.close()
                    inputStream?.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            success
        } else {
            false
        }
    }

    @Synchronized
    private fun copyDataBase(inputStream: InputStream, outputStream: OutputStream): Boolean {
        return try {
            val buffer = ByteArray(1024)
            var length: Int
            while (inputStream.read(buffer).also { length = it } > 0) {
                outputStream.write(buffer, 0, length)
            }
            outputStream.flush()
            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        }
    }

    private fun isValidSQLite(inputFile: Uri): Boolean {
        return try {
            val inputStream = mContentResolver.openInputStream(inputFile)!!
            val fr = InputStreamReader(inputStream)
            val buffer = CharArray(16)
            fr.read(buffer, 0, 16)
            val str = String(buffer)
            fr.close()
            inputStream.close()
            str == "SQLite format 3\u0000"
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    companion object {
        // Backup & Restore
        const val BACKUP_FILE_NAME = "keepdo.db"

        @Volatile
        private var instance: DatabaseAdapter? = null

        fun getInstance(context: Context) = instance ?: synchronized(this) {
            DatabaseAdapter(context).also { instance = it }
        }
    }

}